prov/mrail/src/mrail_cq.c

```diff
@@ -163,6 +142,140 @@ static int mrail_cq_process_comp_buf_recv(struct util_cq *cq,
 	return mrail_cq_process_buf_recv(comp, recv);
 }
 
+static
+struct mrail_early_comp *mrail_peer_get_next(struct mrail_peer_addr *peer_addr)
+{
+	struct slist *queue = &peer_addr->early_recv_comps;
+	struct mrail_early_comp *early_comp;
+
+	if (!slist_empty(queue)) {
+		early_comp = container_of(queue->head, struct
+				mrail_early_comp, entry);
+		if (early_comp->seq_no == peer_addr->expected_seq_no) {
+			slist_remove_head(queue);
+			peer_addr->expected_seq_no++;
```

Comment:
I don't think the locking is correct for this patch.  We probably don't need the locks in mrail_alloc_early_comp() and mrail_free_early_comp() -- because we should likely already be holding one.  There needs to be locks around accessing the expected_seq_no value, and the early_recv_comp queue.

Btw, I would rename early_recv_comp to something like ooo_recv_queue, or at least something with the word queue in the name.